src/middleware.rs

```diff
@@ -63,10 +64,11 @@ impl Chain {
         }
     }
 
-    pub(crate) fn new(logger: slog::Logger) -> Self {
+    pub(crate) fn new_with_stats(logger: slog::Logger, counter: Arc<AtomicUsize>) -> Self {
         Self {
             stack: Arc::new(RwLock::new(vec![
                 Box::new(RetryMiddleware::new(logger)),
+                Box::new(StatsMiddleware::new(counter)),
```

Comment:
I don't love how this counter is passed all over the place. It's created when a `Processor` is created, and then it's passed into the `Chain` type to be incr/decr as work flows through the middleware. It's also used by the `Processor`'s `StatPublisher` to periodically publish the processor metrics. It's just really messy. Any ideas on how to clean this up?

Reply:
To tidy things a little bit I moved the `Arc<AtomicUsize>` into a `Counter` type that will at least keep things easier to read.. the sprawling code is still an issue.